File: docs/examples/nodejs/batch.js
/**
 * Cloudach Batch Inference — Node.js Example
 *
 * Demonstrates:
 *   1. Submitting a batch of chat completion requests
 *   2. Polling for completion
 *   3. Downloading and parsing JSONL results
 *
 * Run:
 *   export CLOUDACH_API_KEY=sk-cloudach-...
 *   node batch.js
 */

const BASE_URL = "https://api.cloudach.com/v1";

const HEADERS = {
  Authorization: `Bearer ${process.env.CLOUDACH_API_KEY}`,
  "Content-Type": "application/json",
};

async function apiFetch(path, { method = "GET", body } = {}) {
  const res = await fetch(`${BASE_URL}${path}`, {
    method,
    headers: HEADERS,
    body: body ? JSON.stringify(body) : undefined,
  });

  if (!res.ok) {
    const data = await res.json().catch(() => ({ error: { message: res.statusText } }));
    throw new Error(data.error?.message ?? res.statusText);
  }
  return res;
}

async function submitBatch(model, requests, completionWindow = "24h") {
  const res = await apiFetch("/batches", {
    method: "POST",
    body: { model, requests, completion_window: completionWindow },
  });
  return res.json();
}

async function getBatch(batchId) {
  const res = await apiFetch(`/batches/${batchId}`);
  return res.json();
}

async function pollUntilDone(batchId, intervalMs = 30_000) {
  while (true) {
    const batch = await getBatch(batchId);
    const { status, progress_pct, completed_count, request_count, failed_count, error } = batch;
    process.stdout.write(
      `  [${status}] ${progress_pct ?? 0}% — ${completed_count}/${request_count} complete` +
        (failed_count ? `, ${failed_count} failed` : "") +
        "\n"
    );

    if (status === "completed") return batch;
    if (["failed", "cancelled", "expired"].includes(status)) {
      throw new Error(`Batch ended with status '${status}'. Error: ${JSON.stringify(error)}`);
    }

    await new Promise((resolve) => setTimeout(resolve, intervalMs));
  }
}

async function downloadResults(batchId) {
  const res = await apiFetch(`/batches/${batchId}/results`);
  const text = await res.text();
  return text
    .trim()
    .split("\n")
    .filter(Boolean)
    .map((line) => JSON.parse(line));
}

// --- Main ---

// Example: sentiment classification on product reviews
const reviews = [
  "Absolutely love this product — works perfectly and arrived early!",
  "Completely broke after two days. Very disappointed.",
  "It's okay, does what it says but nothing special.",
  "Best purchase I've made this year. Highly recommended.",
  "Packaging was damaged and the item smelled strange.",
];

const requests = reviews.map((review, i) => ({
  custom_id: `review-${String(i).padStart(4, "0")}`,
  messages: [
    {
      role: "system",
      content:
        "You are a sentiment classifier. Reply with exactly one word: positive, negative, or neutral.",
    },
    { role: "user", content: review },
  ],
}));

// --- Step 1: Submit ---
console.log(`Submitting batch of ${requests.length} requests...`);
const batch = await submitBatch("llama3-8b", requests, "24h");
const batchId = batch.id;
console.log(`  Batch ID: ${batchId}`);
console.log(`  Estimated cost: $${batch.cost_estimate_usd.toFixed(6)}`);
console.log(`  Estimated completion: ${batch.estimated_completion}`);

// --- Step 2: Poll ---
console.log("\nPolling for completion (30s interval)...");
await pollUntilDone(batchId, 30_000);

// --- Step 3: Download results ---
console.log("\nDownloading results...");
const results = await downloadResults(batchId);
const resultsById = Object.fromEntries(results.map((r) => [r.custom_id, r]));

console.log("\n--- Sentiment Results ---");
reviews.forEach((review, i) => {
  const customId = `review-${String(i).padStart(4, "0")}`;
  const row = resultsById[customId];
  let sentiment;
  if (!row) {
    sentiment = "MISSING";
  } else if (row.error) {
    sentiment = `ERROR: ${row.error.message}`;
  } else {
    sentiment = row.response.choices[0].message.content.trim();
  }
  console.log(`  ${customId}: ${sentiment.padEnd(10)} — ${review.slice(0, 60)}`);
});

File: docs/examples/python/batch.py
"""
Cloudach Batch Inference — Python Example
Requires: pip install requests

Demonstrates:
  1. Submitting a batch of chat completion requests
  2. Polling for completion
  3. Downloading and parsing JSONL results

Run:
  export CLOUDACH_API_KEY=sk-cloudach-...
  python batch.py
"""

import json
import os
import time

import requests

API_KEY = os.environ["CLOUDACH_API_KEY"]
BASE_URL = "https://api.cloudach.com/v1"
HEADERS = {
    "Authorization": f"Bearer {API_KEY}",
    "Content-Type": "application/json",
}


def submit_batch(model: str, requests_payload: list, completion_window: str = "24h") -> dict:
    resp = requests.post(
        f"{BASE_URL}/batches",
        headers=HEADERS,
        json={
            "model": model,
            "requests": requests_payload,
            "completion_window": completion_window,
        },
    )
    resp.raise_for_status()
    return resp.json()


def get_batch(batch_id: str) -> dict:
    resp = requests.get(f"{BASE_URL}/batches/{batch_id}", headers=HEADERS)
    resp.raise_for_status()
    return resp.json()


def poll_until_done(batch_id: str, poll_interval: int = 30) -> dict:
    """Poll until the batch completes. Returns the final batch object."""
    while True:
        batch = get_batch(batch_id)
        status = batch["status"]
        pct = batch.get("progress_pct", 0)
        print(
            f"  [{status}] {pct}% — {batch['completed_count']}/{batch['request_count']} complete"
            + (f", {batch['failed_count']} failed" if batch["failed_count"] else "")
        )

        if status == "completed":
            return batch
        if status in ("failed", "cancelled", "expired"):
            raise RuntimeError(
                f"Batch ended with status '{status}'. Error: {batch.get('error')}"
            )

        time.sleep(poll_interval)


def download_results(batch_id: str) -> list[dict]:
    """Download JSONL results and parse into a list of result objects."""
    resp = requests.get(f"{BASE_URL}/batches/{batch_id}/results", headers=HEADERS)
    resp.raise_for_status()
    return [json.loads(line) for line in resp.text.strip().splitlines()]


def main():
    # --- Build requests ---
    # Example: sentiment classification on product reviews
    reviews = [
        "Absolutely love this product — works perfectly and arrived early!",
        "Completely broke after two days. Very disappointed.",
        "It's okay, does what it says but nothing special.",
        "Best purchase I've made this year. Highly recommended.",
        "Packaging was damaged and the item smelled strange.",
    ]

    requests_payload = [
        {
            "custom_id": f"review-{i:04d}",
            "messages": [
                {
                    "role": "system",
                    "content": "You are a sentiment classifier. Reply with exactly one word: positive, negative, or neutral.",
                },
                {
                    "role": "user",
                    "content": review,
                },
            ],
        }
        for i, review in enumerate(reviews)
    ]

    # --- Submit ---
    print(f"Submitting batch of {len(requests_payload)} requests...")
    batch = submit_batch("llama3-8b", requests_payload, completion_window="24h")
    batch_id = batch["id"]
    print(f"  Batch ID: {batch_id}")
    print(f"  Estimated cost: ${batch['cost_estimate_usd']:.6f}")
    print(f"  Estimated completion: {batch['estimated_completion']}")

    # --- Poll ---
    print("\nPolling for completion (30s interval)...")
    poll_until_done(batch_id, poll_interval=30)

    # --- Download results ---
    print("\nDownloading results...")
    results = download_results(batch_id)

    # Build a lookup by custom_id for easy access
    results_by_id = {r["custom_id"]: r for r in results}

    print("\n--- Sentiment Results ---")
    for i, review in enumerate(reviews):
        custom_id = f"review-{i:04d}"
        row = results_by_id.get(custom_id)
        if row is None:
            sentiment = "MISSING"
        elif row["error"]:
            sentiment = f"ERROR: {row['error']['message']}"
        else:
            sentiment = row["response"]["choices"][0]["message"]["content"].strip()
        print(f"  {custom_id}: {sentiment!r:12} — {review[:60]}")


if __name__ == "__main__":
    main()
